### PR TITLE
Adds patient count API

### PR DESCRIPTION
**Pull request by rkodev**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes https://github.com/google/android-fhir/issues/458

**Description**
Adds a count a part of the search API

**Alternative(s) considered**
Yes, returning a serialized list and getting the size of the list. Introduces unnecessary serialization overhead

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by rkodev** [comment]
@googlebot I signed it!

**Comment by rkodev** [comment]
@jingtang10 @deepankarb @kevinmost @pankajkatoch @yigit This PR is ready for Review

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 1]
```diff
@@ -71,4 +71,12 @@ interface FhirEngine {
   fun updatePeriodicSyncConfiguration(syncConfig: PeriodicSyncConfiguration)
 
   suspend fun <R : Resource> search(search: Search): List<R>
+
+  /**
+   * One time download of resources.
+   *
+   * @param syncConfiguration
+   * - configuration of data that needs to be synchronised
+   */
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/FhirEngine.kt:80]: incorrect doc.

File: engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt [hunk 2]
```diff
@@ -67,7 +71,7 @@ fun Search.getQuery(): SearchQuery {
 
   val query =
     """
-    SELECT a.serializedResource
+    SELECT ${ if (isCount) "count(*)" else "a.serializedResource" }
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/search/MoreSearch.kt:74]:
```suggestion
    SELECT ${ if (isCount) "COUNT(*)" else "a.serializedResource" }
```

File: reference/src/main/java/com/google/android/fhir/reference/PatientListActivity.kt [hunk 3]
```diff
@@ -61,6 +62,11 @@ class PatientListActivity() : AppCompatActivity() {
         adapter.submitList(it)
       }
     )
+
+    patientListViewModel.patientCount.observe(
+      this,
+      { findViewById<TextView>(R.id.patient_count).text = it.toString() + " Patients" }
```
Comment (jingtang10) [inline reference/src/main/java/com/google/android/fhir/reference/PatientListActivity.kt:68]:
```suggestion
      { findViewById<TextView>(R.id.patient_count).text = it.toString() + " Patient(s)" }
```

File: reference/src/main/java/com/google/android/fhir/reference/PatientListViewModel.kt [hunk 4]
```diff
@@ -39,6 +41,19 @@ class PatientListViewModel(application: Application, private val fhirEngine: Fhi
...
+
+  private suspend fun count(): Long {
+    return fhirEngine.count(
+      Search(type = ResourceType.Patient).apply {
+        filter(Patient.ADDRESS_CITY) {
+          prefix = ParamPrefixEnum.EQUAL
+          value = "NAIROBI"
+        }
+      }
+    )
```
Comment (jingtang10) [inline reference/src/main/java/com/google/android/fhir/reference/PatientListViewModel.kt:54]:
```suggestion
    return fhirEngine.count<Patient>(
      filter(Patient.ADDRESS_CITY) {
        prefix = ParamPrefixEnum.EQUAL
        value = "NAIROBI"
      }
    )
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by maimoonak** [comment]
@googlebot I signed it!

**Comment by maimoonak** [comment]
@googlebot I signed it.

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 5]
```diff
@@ -71,6 +71,13 @@ interface FhirEngine {
    * result of the [download] operation.
    */
   suspend fun syncDownload(download: suspend (SyncDownloadContext) -> List<Resource>)
+
+  /**
+   * Total count of entities available for given search
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/FhirEngine.kt:76]: available is misleading. this is the count of resources returned for a given search.

```suggestion
   * Total count of resources for given search
```

**Comment by maimoonak** [comment]
@googlebot I consent.

**Review by jingtang10 — APPROVED** [review]
Thanks! @maimoonak and @rkodev! 👍

**Pull request merged**